fgpyo/alignment/__init__.py

```diff
@@ -0,0 +1,232 @@
+    """
+
+    def __init__(
+        self,
+        query: str,
+        target: str,
+        query_start: int,
+        target_start: int,
+        cigar: Cigar,
+        score: int,
+    ):
+        """
+        Args:
+            query: the query sequence
+            target: the target sequence
+            query_start: the 1-based position in the query sequence where the alignment begins
+            target_start: the 1-based position in the target sequence where the alignment begins
+            cigar: a Cigar object describing the alignment of the two sequences
+            score: the alignment score
+        """
+        self.query: str = query
+        self.target: str = target
+        self.query_start: int = query_start
+        self.target_start: int = target_start
+        self.cigar: Cigar = cigar
+        self.score: int = score
+
+    @property
+    def query_end(self) -> int:
+        """
+        Returns:
+            One based closed coordinate of the end of the alignment on the query sequence.
+        """
+        return self.query_start + self.cigar.length_on_query() - 1
+
+    @property
+    def target_end(self) -> int:
+        """
+        Returns:
+            One based closed coordinate of the end of the alignment on the target sequence.
+        """
+        return self.target_start + self.cigar.length_on_target() - 1
+
+    def padded_string(
+        self,
+        match_char: str = "|",
+        mismatch_char: str = ".",
+        gap_char: str = " ",
+        pad_char: str = "-",
+    ) -> List[str]:
+        """Generates a padded text representation of the alignment for visualization.
+        The returned sequence will consist of three lines as follows (minus the labels on the left):
+
+        query : ACGTGAACTGACT-ACTGTATGCG
+        align : ||||| |||||| ||||||||.|
+        target: ACGTG--CTGACTGACTGTATGGG
+
+        Args:
+            match_char: The character to use in the alignment line when the bases match.
+            mismatch_char: The character to use in the alignment line when the bases mismatch.
+            gap_char: The character to use in the alignment line when one of the
+                sequences is gapped.
+            pad_char: The character to use in the query or target sequence lines when
+                padding is required.
+
+        Returns:
+            Three lines representing the alignment.
+        """
+        query_buffer = []
+        align_buffer = []
+        target_buffer = []
+
+        q_offset = self.query_start - 1
+        t_offset = self.target_start - 1
+
+        for cig in self.cigar.elements:
+            length = cig.length
+
+            if cig.operator == CigarOp.I:
+                for _ in range(length):
+                    query_buffer.append(self.query[q_offset])
+                    align_buffer.append(gap_char)
+                    target_buffer.append(pad_char)
+                    q_offset += 1
```

Comment:
I think you can remove the loops here and in the `CigarOp.D` case
```suggestion
                query_buffer.append(self.query[q_offset: q_offset + length])
                align_buffer.append(gap_char * length)
                target_buffer.append(pad_char * length)
                q_offset += length
```